test/test_autograd.py

```diff
@@ -3240,6 +3240,78 @@ def backward(ctx, grad):
         # check one of them is using the computed buffer
         self.assertTrue(p_a == p_g or p_b == p_g)
 
+    def test_no_grad_copy_sparse(self):
+        # create autograd function that saves grad pointer as class static
+        class MyFunc(Function):
+            static_grad_ptr = None
+
+            @staticmethod
+            def forward(ctx, inp1, inp2):
+                return inp1 + inp2
+
+            @staticmethod
+            def backward(ctx, grad):
+                MyFunc.static_grad_ptr = grad._values().data_ptr()
+                return grad, grad
+
+        class NonContGradFunc(Function):
+            static_grad_ptr = None
+
+            @staticmethod
+            def forward(ctx, inp1, inp2):
+                return inp1 + inp2
+
+            @staticmethod
+            def backward(ctx, grad):
+                # Create a sparse tensor with non-contigous indices and values
+                # and return as grad.
+                v = torch.rand(1, 3)
+                i = torch.ones(1, 1, dtype=torch.long)
+                nv = v.expand(8, 3)
+                ni = i.expand(1, 8)
+                ngrad = torch.sparse.FloatTensor(ni, nv, torch.Size([10, 3]))
+                NonContGradFunc.static_grad_ptr = ngrad._values().data_ptr()
+                return ngrad, ngrad
+
+        a = torch.randn(10, 3, requires_grad=True)
+        b = torch.randn(10, 3, requires_grad=True)
+        input = torch.tensor([1, 2, 4, 5, 4, 3, 2, 9])
+        offsets = torch.tensor([0, 4])
+        import torch.nn.functional as F
+
+        # test case that should trigger no copy for one of a,b
+        emb_matrix = MyFunc.apply(a, b)
+        loss = F.embedding_bag(emb_matrix, input, offsets, sparse=True).sum()
+        loss.backward(retain_graph=True)
+        p_g = MyFunc.static_grad_ptr
+        p_a = a.grad._values().data_ptr()
+        p_b = b.grad._values().data_ptr()
+        # check a,b uses different grad buffer
+        self.assertFalse(p_a == p_b)
+        # check one of them is using the computed buffer
+        self.assertTrue(p_a == p_g or p_b == p_g)
+
+        # Run backwards multiple times to ensure accumulation works.
+        for i in range(100):
+            loss.backward(retain_graph=True)
+
+        # non-contiguous indices and value should not trigger copy either.
```

Comment:
Maybe they should? The current behavior would return contiguous indices and values all the time no?

Reply:
I'm not sure about this, is this a contract we make with users currently? If not it seems like we should optimize this codepath as well. Are there any ops on SparseTensors that fail if indices and values are non-contiguous? Based on what I saw, we usually handle non-contiguous indices and values in SparseTensors appropriately.

Reply:
I'm afraid there is no contract :/ So anything that changes the current behavior is a BC-breaking change.
A user could be giving this directly into a custom cpp function that assumes that indices and values are contiguous for example.